lib/cog/command/pipeline/executor.ex

```diff
@@ -677,35 +663,160 @@ defmodule Cog.Command.Pipeline.Executor do
+    The specific error was:
+
+        #{error_message}
+
+    """
+
+    [base, command_text, error_text]
+    |> Enum.reject(&is_nil/1)
+    |> Enum.join("\n")
   end
 
-  defp send_denied(which, why, state),
-    do: send_reply("Sorry, you aren't allowed to execute '#{which}' :(\n #{why}", state)
+  # Turn an error tuple into a textual message intended for chat users
+  defp format_user_error({:parse_error, msg}, _state) when is_binary(msg),
+    do: "Whoops! An error occurred. " <> msg
+  defp format_user_error({:redirect_error, invalid}, _state),
+    do: "Whoops! An error occurred. " <> redirection_error_message(invalid)
+  defp format_user_error({:binding_error, {:missing_key, var}}, _state),
+    do: "I can't find the variable '$#{var}'."
+  defp format_user_error({:binding_error, msg}, _state) when is_binary(msg),
+    do: "Whoops! An error occurred. " <> msg
+  defp format_user_error({:no_rule, current_invocation}, _state),
+    do: "No rules match the supplied invocation of '#{current_invocation}'. Check your args and options, then confirm that the proper rules are in place."
+  defp format_user_error({:denied, %Piper.Permissions.Ast.Rule{}=rule}, %__MODULE__{invocations: [current_invocation|_]}),
+    do: "Sorry, you aren't allowed to execute '#{current_invocation}' :(\n You will need the '#{rule.permission_selector.perms.value}' permission to run this command."
+  defp format_user_error({:no_relays, %Cog.Models.Bundle{name: name}}, _state),
+    do: "Whoops! An error occurred. No Cog Relays supporting the `#{name}` bundle are currently online"
+  defp format_user_error({:disabled_bundle, %Cog.Models.Bundle{name: name}}, _state),
+    do: "Whoops! An error occurred. The #{name} bundle is currently disabled"
+  defp format_user_error({:timeout, %Cog.Models.Command{}=command}, _state) do
+    name = Cog.Models.Command.full_name(command)
+    "The #{name} command timed out"
+  end
+  defp format_user_error({:template_rendering_error, {error, template, adapter}}, _state),
+    do: "Whoops! An error occurred. There was an error rendering the template '#{template}' for the adapter '#{adapter}': #{inspect error}"
+  defp format_user_error({:command_error, response}, _state),
+    do: "Whoops! An error occurred. " <> (response.status_message || response.body["message"])
 
-  defp send_reply(message, state) do
+  defp send_error(message, state) do
     request = state.request
 
     message = if request["room"]["name"] != "direct" do
-      "@#{request["sender"]["handle"]} " <> message
+      "#{sender_mention_name(state)} #{message}"
     else
       message
     end
     publish_response(message,
                      request["room"],
-                     request["adapter"], # TODO: respect redirects?
+                     request["adapter"],
                      state)
   end
 
+  # Turn an error tuple into a message intended for the audit log
+  defp format_audit_message({:parse_error, msg}, state) when is_binary(msg),
```

Comment:
Possibly... not quite sure at this point yet. I'd probably wait until #469 and #470 are done.

My only concern is about putting the error processing farther away from where it's used (if these were error strings used across multiple modules, that'd be a different story).